libs/ml/include/ml/ops/add.hpp

```diff
@@ -53,21 +53,51 @@ class Add : public fetch::ml::Ops<T>
     assert(inputs.at(0)->shape() == error_signal.shape());
     assert(error_signal.shape() == ComputeOutputShape(inputs));
 
-    // Test if input is broadcastable by batch dimension
-    assert(inputs.at(1)->shape().at(inputs.at(1)->shape().size() - 1) == 1);
-    for (SizeType i{0}; i < inputs.at(0)->shape().size() - 1; i++)
-    {
-      assert(inputs.at(0)->shape().at(i) == inputs.at(1)->shape().at(i));
-    }
-
     if (inputs.at(0)->shape() == inputs.at(1)->shape())
     {
+      // Non-broadcast Add
       return {error_signal, error_signal};
     }
     else
     {
-      SizeType batch_dimension = inputs.at(0)->shape().size() - 1;
-      return {error_signal, fetch::math::ReduceSum(error_signal, batch_dimension)};
+      // Broadcast Add
```

Comment:
please move lines 63 - 98 into a separate private member function since it's all just code to update axes.
It feels like a function that will get reused by a lot of ops, so maybe we make it a more general utility in the future?